res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js

```diff
@@ -0,0 +1,978 @@
+    this.clipLeftConnection = engine.makeConnection("[Channel1]", "PeakIndicator", this.peakOutputHandler);
+    this.clipRightConnection = engine.makeConnection("[Channel2]", "PeakIndicator", this.peakOutputHandler);
+
+    // Sampler callbacks
+    for (i = 1; i <= 16; ++i) {
+        this.samplerCallbacks.push(engine.makeConnection("[Sampler" + i + "]", "track_loaded", this.samplesOutputHandler));
+        this.samplerCallbacks.push(engine.makeConnection("[Sampler" + i + "]", "play_indicator", this.samplesOutputHandler));
+    }
+
+    TraktorS2MK3.lightDeck(false);
+};
+
+/* Helper function to link output in a short form */
+TraktorS2MK3.linkOutput = function (group, name, callback) {
+    TraktorS2MK3.controller.linkOutput(group, name, group, name, callback);
+};
+
+TraktorS2MK3.vuMeterHandler = function (value, group, key) {
+    var vuKeys = Object.keys(TraktorS2MK3.vuMeterThresholds);
+    for (var i = 0; i < vuKeys.length; ++i) {
+        // Avoid spamming HID by only sending last LED update
+        var last = (i === (vuKeys.length - 1));
+        if (TraktorS2MK3.vuMeterThresholds[vuKeys[i]] > value) {
+            TraktorS2MK3.controller.setOutput(group, vuKeys[i], 0x00, last);
+        } else {
+            TraktorS2MK3.controller.setOutput(group, vuKeys[i], 0x7E, last);
+        }
+    }
+};
+
+TraktorS2MK3.peakOutputHandler = function (value, group, key) {
+    var ledValue = 0x00;
+    if (value) {
+        ledValue = 0x7E;
+    }
+
+    TraktorS2MK3.controller.setOutput(group, key, ledValue, true);
+};
+
+TraktorS2MK3.outputHandler = function (value, group, key) {
+    // Custom value for multi-colored LEDs
+    var ledValue = value;
+    if (value === 0 || value === false) {
+        // Off value
+        ledValue = 0x7C;
+    } else if (value === 1 || value === true) {
+        // On value
+        ledValue = 0x7E;
+    }
+
+    TraktorS2MK3.controller.setOutput(group, key, ledValue, true);
+};
+
+TraktorS2MK3.hotcueOutputHandler = function (value, group, key) {
+    // Light button LED only when we are in hotcue mode
+    if (TraktorS2MK3.padModeState[group] === 0) {
+        TraktorS2MK3.outputHandler(value, group, key);
+    }
+};
+
+TraktorS2MK3.samplesOutputHandler = function (value, group, key) {
+    // Sampler 1-4, 9-12 -> Channel1
+    // Samples 5-8, 13-16 -> Channel2
+    var sampler = TraktorS2MK3.resolveSampler(group);
+    var deck = "[Channel1]";
+    var num = sampler;
+    if (sampler === undefined) {
+        return;
+    } else if (sampler > 4 && sampler < 9) {
+        deck = "[Channel2]";
+        num = sampler - 4;
+    } else if (sampler > 8 && sampler < 13) {
+        num = sampler - 4;
+    } else if (sampler > 12 && sampler < 17) {
+        deck = "[Channel2]";
+        num = sampler - 8;
+    }
+
+    // If we are in samples modes light corresponding LED
+    if (TraktorS2MK3.padModeState[deck] === 1) {
+        if (key === "play_indicator") {
```

Comment:
Use `play`, not `play_indicator` here

Reply:
I use `play` now in https://github.com/mi01/mixxx/commit/3f0fd915ae129993968be79bcb867eb8f22bb468

But i discovered an issue here. In this callback function i process events of type `play` and `track_loaded`. When a track is ejected from a sampler slot, i receive an event of type `track_loaded`. After this i receive immediately two times an event of type `play` with an value of 0 for the same sampler slot, even though no track is loaded here anymore. This is very strange and i could not figure out so far, if this is a problem in my script or of Mixxx.

Reply:
Hmm, that seems like a bug in Mixxx. Does this script still work as intended? From reading the code it doesn't seem to me that the redundant event causes a problem for this script.

Reply:
The script still works because i check `engine.getValue(group, "track_loaded")` before processing the `play` event now.

Reply:
Okay. I wonder if this double `play` event occurs with normal decks too or just samplers?

Reply:
I tested it right now and it seems to happen also on `[ChannelX]`.

Reply:
Thanks for testing. I reported a [bug](https://bugs.launchpad.net/mixxx/+bug/1856358)